server/lib/src/valueset/image.rs

```diff
@@ -0,0 +1,663 @@
+static PNG_CHUNK_END: &[u8; 4] = b"IEND";
+
+pub trait ImageValueThings {
+    fn validate_image(&self) -> Result<(), ImageValidationError>;
+    fn validate_is_png(&self) -> Result<(), ImageValidationError>;
+    fn validate_is_gif(&self) -> Result<(), ImageValidationError>;
+    fn validate_is_jpg(&self) -> Result<(), ImageValidationError>;
+    fn validate_is_webp(&self) -> Result<(), ImageValidationError>;
+    fn validate_is_svg(&self) -> Result<(), ImageValidationError>;
+
+    /// A sha256 of the filename/type/contents
+    fn hash_imagevalue(&self) -> String;
+
+    fn get_limits(&self) -> image::io::Limits {
+        let mut limits = image::io::Limits::default();
+        limits.max_image_height = Some(MAX_IMAGE_HEIGHT);
+        limits.max_image_width = Some(MAX_IMAGE_WIDTH);
+        limits
+    }
+}
+
+#[derive(Debug)]
+pub enum ImageValidationError {
+    Acropalypse(String),
+    ExceedsMaxWidth,
+    ExceedsMaxHeight,
+    ExceedsMaxDimensions,
+    ExceedsMaxFileSize,
+    InvalidImage(String),
+    InvalidPngPrelude,
+}
+
+impl Display for ImageValidationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ImageValidationError::ExceedsMaxWidth => f.write_fmt(format_args!(
+                "Exceeds the maximum width: {}",
+                MAX_IMAGE_WIDTH
+            )),
+            ImageValidationError::ExceedsMaxHeight => f.write_fmt(format_args!(
+                "Exceeds the maximum height: {}",
+                MAX_IMAGE_HEIGHT
+            )),
+            ImageValidationError::ExceedsMaxFileSize => f.write_fmt(format_args!(
+                "Exceeds maximum file size of {}",
+                MAX_FILE_SIZE
+            )),
+            ImageValidationError::InvalidImage(message) => {
+                if !message.is_empty() {
+                    f.write_fmt(format_args!("Invalid Image: {}", message))
+                } else {
+                    f.write_str("Invalid Image")
+                }
+            }
+            ImageValidationError::ExceedsMaxDimensions => f.write_fmt(format_args!(
+                "Image exceeds max dimensions of {}x{}",
+                MAX_IMAGE_WIDTH, MAX_IMAGE_HEIGHT
+            )),
+            ImageValidationError::Acropalypse(message) => {
+                if !message.is_empty() {
+                    f.write_fmt(format_args!(
+                        "Image has extra data, is vulnerable to Acropalypse: {}",
+                        message
+                    ))
+                } else {
+                    f.write_str("Image has extra data, is vulnerable to Acropalypse")
+                }
+            }
+            ImageValidationError::InvalidPngPrelude => {
+                f.write_str("Image has an invalid PNG prelude and is likely corrupt.")
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ChunkStatus {
+    SeenEnd { has_trailer: bool },
+    MoreChunks,
+}
+#[test]
```

Comment:
Should this test be at the foot of the file? Seems messy up here ...

Reply:
'tis now, whoops. :smile: